epochStart/metachain/auctionListSelector.go

```diff
@@ -0,0 +1,255 @@
+	ShardCoordinator             sharding.Coordinator
+	StakingDataProvider          epochStart.StakingDataProvider
+	MaxNodesChangeConfigProvider epochStart.MaxNodesChangeConfigProvider
+}
+
+// NewAuctionListSelector will create a new auctionListSelector, which handles selection of nodes from auction list based
+// on their top up
+func NewAuctionListSelector(args AuctionListSelectorArgs) (*auctionListSelector, error) {
+	if check.IfNil(args.ShardCoordinator) {
+		return nil, epochStart.ErrNilShardCoordinator
+	}
+	if check.IfNil(args.StakingDataProvider) {
+		return nil, epochStart.ErrNilStakingDataProvider
+	}
+	if check.IfNil(args.MaxNodesChangeConfigProvider) {
+		return nil, epochStart.ErrNilMaxNodesChangeConfigProvider
+	}
+
+	asl := &auctionListSelector{
+		shardCoordinator:    args.ShardCoordinator,
+		stakingDataProvider: args.StakingDataProvider,
+		nodesConfigProvider: args.MaxNodesChangeConfigProvider,
+	}
+
+	return asl, nil
+}
+
+// SelectNodesFromAuctionList will select nodes from validatorsInfoMap based on their top up. If two or more validators
+// have the same top-up, then sorting will be done based on blsKey XOR randomness. Selected nodes will have their list set
+// to common.SelectNodesFromAuctionList
+func (als *auctionListSelector) SelectNodesFromAuctionList(validatorsInfoMap state.ShardValidatorsInfoMapHandler, randomness []byte) error {
+	if len(randomness) == 0 {
+		return process.ErrNilRandSeed
+	}
+
+	currNodesConfig := als.nodesConfigProvider.GetCurrentNodesConfig()
+	numOfShuffledNodes := currNodesConfig.NodesToShufflePerShard * (als.shardCoordinator.NumberOfShards() + 1)
+
+	auctionList, currNumOfValidators := getAuctionListAndNumOfValidators(validatorsInfoMap)
+	numOfValidatorsAfterShuffling, err := safeSub(currNumOfValidators, numOfShuffledNodes)
+	if err != nil {
+		log.Warn(fmt.Sprintf("%v when trying to compute numOfValidatorsAfterShuffling = %v - %v (currNumOfValidators - numOfShuffledNodes)",
+			err,
+			currNumOfValidators,
+			numOfShuffledNodes,
+		))
+		numOfValidatorsAfterShuffling = 0
+	}
+
+	maxNumNodes := currNodesConfig.MaxNumNodes
+	availableSlots, err := safeSub(maxNumNodes, numOfValidatorsAfterShuffling)
+	if availableSlots == 0 || err != nil {
+		log.Info(fmt.Sprintf("%v or zero value when trying to compute availableSlots = %v - %v (maxNodes - numOfValidatorsAfterShuffling); skip selecting nodes from auction list",
+			err,
+			maxNumNodes,
+			numOfValidatorsAfterShuffling,
+		))
+		return nil
+	}
+
+	auctionListSize := uint32(len(auctionList))
+	log.Info("systemSCProcessor.SelectNodesFromAuctionList",
+		"max nodes", maxNumNodes,
+		"current number of validators", currNumOfValidators,
+		"num of nodes which will be shuffled out", numOfShuffledNodes,
+		"num of validators after shuffling", numOfValidatorsAfterShuffling,
+		"auction list size", auctionListSize,
+		fmt.Sprintf("available slots (%v -%v)", maxNumNodes, numOfValidatorsAfterShuffling), availableSlots,
```

Comment:
> %v - %v*

Any particular reason for this suggestion? I like the current version better

Reply:
I think `available slots (3 -5)` is better than `available slots (3 - 5)`.
Also, why don't you use `%d` instead of `%v` ?

Reply:
>Because available slots (3 - 5) is better than available slots (3 -5)

Right, I thought you suggested to remove the paranthesis

>Also, why don't you use %d instead of %v ?

`%v` is the default value in golang, and for decimals it translates to `%d`, so it should make no difference, right?